File: src/test/java/dev/jiaming/ai_interview/rag/RagIndexCleanupSchedulerTests.java
package dev.jiaming.ai_interview.rag;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import java.time.Duration;

import org.junit.jupiter.api.Test;

import dev.jiaming.ai_interview.common.RuntimeModeProperties;

class RagIndexCleanupSchedulerTests {

	private static final Duration RETENTION = Duration.ofDays(7);

	private final RagIndexingService indexingService = mock(RagIndexingService.class);

	@Test
	void runsCleanupWhenTheWorkerIsEnabled() {
		RagIndexCleanupScheduler scheduler = new RagIndexCleanupScheduler(
			indexingService, new RuntimeModeProperties("worker")
		);

		scheduler.cleanup();

		verify(indexingService).cleanupStale(RETENTION, 100);
	}

	@Test
	void runsCleanupInCombinedMode() {
		RagIndexCleanupScheduler scheduler = new RagIndexCleanupScheduler(
			indexingService, new RuntimeModeProperties("all")
		);

		scheduler.cleanup();

		verify(indexingService).cleanupStale(RETENTION, 100);
	}

	@Test
	void skipsCleanupInApiOnlyMode() {
		RagIndexCleanupScheduler scheduler = new RagIndexCleanupScheduler(
			indexingService, new RuntimeModeProperties("api")
		);

		scheduler.cleanup();

		verify(indexingService, never()).cleanupStale(any(), anyInt());
	}
}

File: src/test/kotlin/dev/jiaming/ai_interview/rag/RagIndexCleanupSchedulerTests.kt
package dev.jiaming.ai_interview.rag

import dev.jiaming.ai_interview.common.RuntimeModeProperties
import java.time.Duration
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mockito

class RagIndexCleanupSchedulerTests {
	private val indexingService = Mockito.mock(RagIndexingService::class.java)

	@Test
	fun runsCleanupWhenTheWorkerIsEnabled() {
		RagIndexCleanupScheduler(indexingService, RuntimeModeProperties("worker")).cleanup()
		Mockito.verify(indexingService).cleanupStale(RETENTION, 100)
	}

	@Test
	fun runsCleanupInCombinedMode() {
		RagIndexCleanupScheduler(indexingService, RuntimeModeProperties("all")).cleanup()
		Mockito.verify(indexingService).cleanupStale(RETENTION, 100)
	}

	@Test
	fun skipsCleanupInApiOnlyMode() {
		RagIndexCleanupScheduler(indexingService, RuntimeModeProperties("api")).cleanup()
		Mockito.verify(indexingService, Mockito.never()).cleanupStale(any(), anyInt())
	}

	private companion object {
		val RETENTION: Duration = Duration.ofDays(7)
	}
}
